// @onramper/moonpay-adapter/build/KYC/items.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.profession = exports.netWorth = exports.annualExpectedActivity = exports.sourceOfFunds = exports.grossAnnualIncome = exports.employmentStatus = exports.accountPurpose = exports.cryptocurrencyAddressTag = exports.creditCardTokenId = exports.transactionId = exports.cryptocurrencyAddress = exports.bankAccountNumberItem = exports.bankSortCodeItem = exports.bankIbanItem = exports.verifyPhoneCodeItem = exports.phoneCountryCodeItem = exports.phoneNumberItem = exports.verifyEmailCodeItem = exports.emailItem = exports.optionalStateItem = exports.stateItem = exports.countryItem = exports.postCodeItem = exports.townItem = exports.streetItem = exports.lastNameItem = exports.firstNameItem = exports.dateOfBirthItem = void 0;
exports.dateOfBirthItem = {
    type: "date",
    name: "dateOfBirth",
    humanName: "Date of Birth",
    data: [
        {
            type: "integer",
            humanName: "Day",
            name: "day",
        },
        {
            type: "integer",
            humanName: "Month",
            name: "month",
        },
        {
            type: "integer",
            humanName: "Year",
            name: "year",
        },
    ],
};
exports.firstNameItem = {
    type: "string",
    name: "firstName",
    humanName: "First name",
};
exports.lastNameItem = {
    type: "string",
    name: "lastName",
    humanName: "Last Name",
};
exports.streetItem = {
    type: "string",
    name: "street",
    humanName: "Street",
};
exports.townItem = {
    type: "string",
    name: "town",
    humanName: "City",
};
exports.postCodeItem = {
    type: "string",
    name: "postCode",
    humanName: "Postal Code",
};
exports.countryItem = {
    type: "string",
    name: "country",
    humanName: "Country",
};
exports.stateItem = {
    type: "string",
    name: "state",
    humanName: "State",
};
exports.optionalStateItem = Object.assign(Object.assign({}, exports.stateItem), { hint: "Only required if you ", required: false });
exports.emailItem = {
    type: "string",
    name: "email",
    humanName: "Email",
    hint: "We will send a code to your email.",
};
exports.verifyEmailCodeItem = {
    type: "string",
    name: "verifyEmailCode",
    humanName: "Email verification code",
};
exports.phoneNumberItem = {
    type: "integer",
    name: "phoneNumber",
    humanName: "Phone number",
};
exports.phoneCountryCodeItem = {
    type: "integer",
    name: "phoneCountryCode",
    humanName: "Country code",
};
exports.verifyPhoneCodeItem = {
    type: "string",
    name: "verifyPhoneCode",
    humanName: "Phone verification code",
};
exports.bankIbanItem = {
    type: "string",
    name: "bankIban",
    humanName: "IBAN",
};
exports.bankSortCodeItem = {
    type: "string",
    name: "bankSortCode",
    humanName: "Sort code",
};
exports.bankAccountNumberItem = {
    type: "string",
    name: "bankAccountNumber",
    humanName: "Bank account number",
};
exports.cryptocurrencyAddress = {
    type: "string",
    name: "cryptocurrencyAddress",
    humanName: "Cryptocurrency wallet address",
};
exports.transactionId = {
    type: "string",
    name: "transactionId",
    humanName: "Transaction Id",
};
exports.creditCardTokenId = {
    type: "string",
    name: "ccTokenId",
    humanName: "Credit Card Token Id",
};
exports.cryptocurrencyAddressTag = {
    type: "string",
    name: "cryptocurrencyAddressTag",
    humanName: "Cryptocurrency address tag",
    required: false,
};
exports.accountPurpose = {
    type: "select",
    name: "accountPurpose",
    humanName: "How do you intend to use Moonpay?",
    required: true,
    options: [
        {
            value: "Beginner",
            humanName: "Beginner seeking investment",
        },
        {
            value: "Trading",
            humanName: "Trading",
        },
        {
            value: "Payments",
            humanName: "Making payments",
        },
        {
            value: "Other",
            humanName: "Other",
        },
    ],
};
exports.employmentStatus = {
    type: "select",
    name: "employmentStatus",
    humanName: "What is your current employment status?",
    required: true,
    options: [
        {
            value: "Employed",
            humanName: "Employed",
        },
        {
            value: "SelfEmployed",
            humanName: "Self Employed",
        },
        {
            value: "Student",
            humanName: "Student",
        },
        {
            value: "NotEmployed",
            humanName: "Not Employed",
        },
        {
            value: "Retired",
            humanName: "Retired",
        },
    ],
};
exports.grossAnnualIncome = {
    type: "select",
    name: "grossAnnualIncome",
    humanName: "What is your yearly gross income in EUR?",
    required: true,
    options: [
        {
            value: "LessThan10k",
            humanName: "9,999 EUR or less",
        },
        {
            value: "MoreThan10k",
            humanName: "10,000 - 19,999 EUR",
        },
        {
            value: "MoreThan20k",
            humanName: "20,000 - 39,999 EUR",
        },
        {
            value: "MoreThan40k",
            humanName: "40,000 - 59,999 EUR",
        },
        {
            value: "MoreThan60k",
            humanName: "60,000 - 79,999 EUR",
        },
        {
            value: "MoreThan80k",
            humanName: "80,000 - 99,999 EUR",
        },
        {
            value: "MoreThan100k",
            humanName: "100,000 EUR or more",
        },
    ],
};
exports.sourceOfFunds = {
    type: "select",
    name: "sourceOfFunds",
    humanName: "What is the origin of funds?",
    required: true,
    options: [
        {
            value: "Salary",
            humanName: "Salary (Personal income)",
        },
        {
            value: "Savings",
            humanName: "Savings",
        },
        {
            value: "Pension",
            humanName: "Pension",
        },
        {
            value: "Investments",
            humanName: "Income from investments",
        },
        {
            value: "Other",
            humanName: "Other",
        },
        {
            value: "Mining",
            humanName: "Mining",
        },
    ],
};
exports.annualExpectedActivity = {
    type: "select",
    name: "annualExpectedActivity",
    humanName: "What is your annual expected activity on Moonpay?",
    required: true,
    options: [
        {
            value: "LessThan1k",
            humanName: "999 EUR or less",
        },
        {
            value: "MoreThan1k",
            humanName: "1,000 - 9,999 EUR",
        },
        {
            value: "MoreThan10k",
            humanName: "10,000 - 24,999 EUR",
        },
        {
            value: "MoreThan25k",
            humanName: "25,000 EUR or more",
        },
    ],
};
exports.netWorth = {
    type: "select",
    name: "netWorth",
    humanName: "What is your rough net worth in EUR?",
    required: true,
    options: [
        {
            value: "LessThan50k",
            humanName: "49,999 EUR or less",
        },
        {
            value: "MoreThan50k",
            humanName: "50,000 - 99,000 EUR",
        },
        {
            value: "MoreThan100k",
            humanName: "100,000 - 499,000 EUR",
        },
        {
            value: "MoreThan500k",
            humanName: "1,000,000 EUR or more",
        },
    ],
};
exports.profession = {
    type: "string",
    name: "profession",
    humanName: "What is your profession?",
    required: true,
};
