// @onramper/moonpay-adapter/build/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.baseUploadsUrl = exports.baseAPIUrl = exports.baseCreditCardSandboxUrl = exports.acceptedContentTypes = exports.identifier = exports.moonpayBaseAPI = exports.publishableApiKey = void 0;
function publishableApiKey(apiKey) {
    return apiKey.substr("pk_".length, "prod".length) === "prod"
        ? "pk_live_MJouyH3Fe3BVaNOaHCt5SvBfJ5Mk7RvQ"
        : "pk_test_PjABKr88VlgosyTueq3exrVnYYLd4ZB";
}
exports.publishableApiKey = publishableApiKey;
exports.moonpayBaseAPI = "https://api.moonpay.io/v3";
// This identifier is also referenced by copy in `infra/src/lambda-edge.ts`, any changes here should also be applied there
exports.identifier = "Moonpay";
exports.acceptedContentTypes = [
    "image/jpeg",
    "image/png",
    "application/pdf",
];
const domain = process.env.STAGE === "prod" ? "onramper.tech" : "staging.onramper.tech";
exports.baseCreditCardSandboxUrl = `https://moonpay.sandbox.${domain}`;
exports.baseAPIUrl = `https://${domain}`;
exports.baseUploadsUrl = `https://upload.${domain}`;
